### ISSUE TITLE
Notebook LM Voices

### ISSUE THREAD

**Issue by uetuluk** [davidbrowne17/csm-streaming#20]
Is it feasible to use a podcast to train CSM to recognize 2 speakers talking to each other to create a local notebook LM that sounds nice?

**Comment by uetuluk** [comment]
I have taken a look and it seems like a hack might be possible since Llama can generate multiple people talking by default, not sure what will be the result. 

Below is what o4-mini-high thinks is the solution, leaving it in case it inspires anyone.



You’ve actually already built most of the plumbing for multi‑speaker support—your `AudioTextPair` carries a `speaker_id` and your tokenizer will inject that ID into the text.  All you need to do is:

1. **Rip out the global `SPEAKER_ID` constant**, and load each file with its own ID.  
2. **Feed that ID all the way through your `Dataset → collate_fn → model`**, so you can condition on it.  
3. (Optionally) **Add an explicit speaker‑embedding layer** in your model for extra capacity.

---

### 1. Remove the hard‑coded SPEAKER_ID

```diff
- SPEAKER_ID = 0
```

Instead, when you create your `AudioTextPair`s, assign each one its true `speaker_id` (e.g. by parsing it out of the filename or a CSV/JSON metadata file):

```python
def transcribe_audio_files():
    # …  
    for audio_file in tqdm(audio_files):
        # parse “speaker3_hello.wav” → speaker_id = 3
        basename = os.path.basename(audio_file)
        speaker_id = int(basename.split("_")[0].replace("speaker", ""))
        result = transcriber(audio_file)
        transcription = result["text"].strip()
        audio_text_pairs.append(
            AudioTextPair(audio_path=audio_file,
                          text=transcription,
-                         speaker_id=0)
+                         speaker_id=speaker_id)
        )
```

---

### 2. Pass `speaker_id` through the `Dataset` and `collate_fn`

In your `CSMDataset.__getitem__`, you already do:

```python
text_tokens, text_masks = self.tokenize_text_segment(item.text, item.speaker_id)
```

but your `collate_fn` needs to carry those IDs into the batch:

```python
def collate_fn(batch):
    device = batch[0]["input_tokens"].device
    # … your existing padding logic …
    speaker_ids = torch.tensor(
        [item["speaker_id"] for item in batch],
        dtype=torch.long,
        device=device
    )
    return {
        "input_tokens": …,
        "input_masks": …,
        "target_tokens": …,
        "target_masks": …,
        "positions": …,
+       "speaker_ids": speaker_ids
    }
```

And in your training loop, unpack:

```python
for step, batch in enumerate(dataloader):
    speaker_ids = batch.pop("speaker_ids")
    # …
    loss = single_pass_forward(model, bridging_module,
                               target_tokens, target_masks,
                               positions, speaker_ids)
```

---

### 3. (Optional) Add a speaker‑embedding in the model

If you want the model to learn, say, a 64‑dim “voice identity” embedding rather than just relying on the “[3]” token in the text stream, you can:

1.  In your `prepare_csm_model_for_training`, after loading `model`:
    ```python
    NUM_SPEAKERS = 10          # however many you have
    SPEAKER_EMB_DIM = 64
    model.speaker_embedding = nn.Embedding(NUM_SPEAKERS, SPEAKER_EMB_DIM).to(DEVICE)
    ```

2.  In `single_pass_forward`, before you call `model._embed_tokens`:
    ```python
    # batch_size, seq_len
    bsz, seq_len = target_tokens.size()[:2]
    # expand (b,) → (b, seq_len, dim)
    spk_emb = model.speaker_embedding(speaker_ids)          # (b, dim)
    spk_emb = spk_emb.unsqueeze(1).expand(-1, seq_len, -1)  # (b, seq_len, dim)
    ```

3.  And then inject it into your token embedding:
    ```python
    embed = model._embed_tokens(target_tokens)   # (b, seq_len, D)
    embed = embed + spk_emb                     # broadcast-add speaker info
    masked_embed = embed * target_masks.unsqueeze(-1)
    ```

---

With those three changes:

- **Each file** carries its own `speaker_id`
- Your **dataset** and **collate_fn** preserve that ID in the batch
- Your **forward pass** can now condition on which voice is speaking

…and your model will learn to handle an arbitrary number of different speakers (and even conversations between them) in a single training run.
